hi-ml/src/health_ml/utils/regression_test_utils.py

```diff
@@ -0,0 +1,169 @@
+    :param csv_relative_tolerance: When comparing CSV files, use this as the maximum allowed relative discrepancy.
+    If 0.0, do not allow any discrepancy.
+    :return: A list of human readable error messages, with message and file path. If no errors are found, the list is
+    empty.
+    """
+    messages = []
+    if run and not is_running_in_azure_ml(run):
+        logging.warning("Skipping file comparison because the given run context is an AzureML offline run.")
+        return []
+    files_in_run: List[str] = run.get_file_names() if run else []
+    temp_folder = Path(tempfile.mkdtemp()) if run else None
+    for file in expected_folder.rglob("*"):
+        # rglob also returns folders, skip those
+        if file.is_dir():
+            continue
+        # All files stored in AzureML runs use Linux-style path
+        file_relative = file.relative_to(expected_folder).as_posix()
+        if actual_folder:
+            actual_file = actual_folder / file_relative
+        elif temp_folder is not None and run is not None:
+            actual_file = temp_folder / file_relative
+            if file_relative in files_in_run:
+                run.download_file(name=str(file_relative), output_file_path=str(actual_file))
+        else:
+            raise ValueError("One of the two arguments run, actual_folder must be provided.")
+        message = compare_files(expected=file, actual=actual_file,
+                                csv_relative_tolerance=csv_relative_tolerance) if actual_file.exists() else MISSING_FILE
+        if message:
+            messages.append(f"{message}: {file_relative}")
+        logging.info(f"File {file_relative}: {message or 'OK'}")
+    if temp_folder:
+        shutil.rmtree(temp_folder)
+    return messages
+
+
+def compare_folders_and_run_outputs(expected: Path, actual: Path, csv_relative_tolerance: float) -> None:
+    """
+    Compares the actual set of run outputs in the `actual` folder against an expected set of files in the `expected`
+    folder. The `expected` folder can have two special subfolders AZUREML_OUTPUT and AZUREML_PARENT_OUTPUT, that
+    contain files that are expected to be present in the AzureML run context of the present run (AZUREML_OUTPUT)
+    or the run context of the parent run (AZUREML_PARENT_OUTPUT).
+    If a file is missing, or does not have the expected contents, an exception is raised.
+
+    :param expected: A folder with files that are expected to be present.
+    :param actual: The output folder with the actually produced files.
+    :param csv_relative_tolerance: When comparing CSV files, use this as the maximum allowed relative discrepancy.
+    If 0.0, do not allow any discrepancy.
+    """
+    if not expected.is_dir():
+        raise ValueError(f"Folder with expected files does not exist: {expected}")
+    logging.debug(f"Current working directory: {Path.cwd()}")
+    messages = []
+    for (subfolder, message_prefix, actual_folder, run_to_compare) in \
```

Comment:
Can we create the iterable first and use `zip` in the for loop to improve readability?

Reply:
I tried a zip solution, but I don't really think it improves readability. Have now at least taken the list out of the `for` statement